src/components/ContextMenu/ContextMenu.jsx

```diff
@@ -0,0 +1,228 @@
+					this.props.onClickOut(event);
+				}
+			}
+		});
+	},
+
+	componentWillUnmount() {
+		clearInterval(this.updateTargetRectangleIntervalId);
+		window.document.body.removeEventListener('click', this.onClickBodyEventListener);
+	},
+
+	componentWillReceiveProps() {
+		let targetDOMNode = this.refs.target;
+
+		if (targetDOMNode) {
+			this.alignFlyOut(targetDOMNode);
+		}
+	},
+
+	statics: {
+		DOWN: 'down',
+		UP: 'up',
+		BELOW_FOLD: 'BELOW_FOLD',
+		ABOVE_FOLD: 'ABOVE_FOLD',
+		isElementBelowFold(targetElement, yPosition) {
+			if (typeof window !== 'undefined') {
+				let bodyRect = window.document.body.getBoundingClientRect();
+				let targetElementRect = targetElement.getBoundingClientRect();
+				yPosition = (yPosition === undefined ? targetElementRect.top : yPosition);
+				if (yPosition + targetElementRect.height - bodyRect.top > window.innerHeight - bodyRect.top) {
+					return true;
+				} else {
+					return false;
+				}
+			} else {
+				throw new Error('Browser only! Cannot access window object.');
+			}
+		}
+	},
+
+	alignFlyOut(targetDOMNode) {
+		let flyOutHeight = this.state.flyOutHeight;
+		if (this.refs.flyOutPortal) {
+			let flyOutEl = this.refs.flyOutPortal.portalElement.firstChild;
+			flyOutHeight = flyOutEl.getBoundingClientRect().height;
+
+			if (this.props.onChangeBounds) {
+				let isFlyoutBelowFold = ContextMenu.isElementBelowFold(flyOutEl, targetDOMNode.getBoundingClientRect().bottom);
+				if (this.state.isFlyoutBelowFold !== isFlyoutBelowFold) {
+					this.setState({
+						isFlyoutBelowFold: isFlyoutBelowFold
+					});
+					this.props.onChangeBounds(isFlyoutBelowFold ? ContextMenu.BELOW_FOLD : ContextMenu.ABOVE_FOLD);
+				}
+			}
+		}
+
+		this.setState({
+			targetRect: getAbsoluteBoundingClientRect(targetDOMNode),
+			flyOutHeight: flyOutHeight
+		});
+	},
+
+	render() {
+		const {
+			className,
+			style,
+			isExpanded,
+			direction
+		} = this.props;
+
+		const {
+			portalId,
+			targetRect,
+			flyOutHeight
+		} = this.state;
+
+		const targetProps = _.first(ContextMenu.Target.findInAllAsProps(this.props));
+		const targetChildren = targetProps.children;
+		const flyProps = _.first(ContextMenu.FlyOut.findInAllAsProps(this.props));
+		const flyChildren = flyProps.children;
+
+
+		return (
+			<span ref='target' className={classNames(boundClassNames('~'), className)} style={style}>
+				{targetChildren}
+				{isExpanded ? (
+					<Portal
+						ref='flyOutPortal'
+						{...flyProps}
+						className={classNames(boundClassNames('FlyOut', {
+							'FlyOut-Up': direction === ContextMenu.UP,
+							'FlyOut-Down': direction === ContextMenu.DOWN
+						}), flyProps.className)}
+						portalId={portalId}
+						style={_.assign({}, flyProps.style, {
+							position: 'absolute',
+							left: targetRect.left,
+							minWidth: targetRect.width
+						}, (direction === ContextMenu.UP ? {
+							top: targetRect.top - flyOutHeight
+						} : {
+							top: targetRect.bottom
+						}))}
```

Comment:
if i do the JSX object spread here it will overwrite my essential styles, which are needed for this component to render properly. I think this is one of the few cases where i need to assign/merge style from props with the styles.